plugins/repository-gcs/src/main/java/io/crate/gcs/GCSBlobContainer.java

```diff
@@ -0,0 +1,224 @@
+        String prefix = path().buildAsString();
+        Map<String, BlobContainer> result = new HashMap<>();
+        bucket.list(Storage.BlobListOption.currentDirectory(), Storage.BlobListOption.prefix(prefix))
+            .iterateAll()
+            .forEach(blob -> {
+                if (blob.isDirectory()) {
+                    assert blob.getName().startsWith(prefix);
+                    assert blob.getName().endsWith("/");
+                    final String suffixName = blob.getName().substring(prefix.length(), blob.getName().length() - 1);
+                    if (suffixName.isEmpty() == false) {
+                        result.put(suffixName, new GCSBlobContainer(bucket, path().add(suffixName)));
+                    }
+                }
+            });
+        return result;
+    }
+
+    @Override
+    public Map<String, BlobMetadata> listBlobs() {
+        return listBlobsByPrefix("");
+    }
+
+    @Override
+    public Map<String, BlobMetadata> listBlobsByPrefix(String blobNamePrefix) {
+        final String basePath = buildKey(blobNamePrefix);
+        final List<Blob> blobs = findBlobsByPrefix(basePath);
+        return convertBlobsToMap(normalizeBlobs(basePath, blobs));
+    }
+
+    /**
+     * Create a new empty blob with the specified name.
+     */
+    private Blob createBlob(String blobName, boolean failIfAlreadyExists) throws IOException {
+        String key = buildKey(blobName);
+        final Blob blob = bucket.get(key);
+        if (blob != null) {
+            if (failIfAlreadyExists) {
+                throw new IOException("blob `" + blobName + "` already exists");
+            } else {
+                return blob;
+            }
+        }
+        return bucket.create(key, new byte[]{});
+    }
+
+    /**
+     * List all blobs starting with a given prefix.
+     */
+    private List<Blob> findBlobsByPrefix(String blobNamePrefix) {
+        final Page<Blob> blobs = bucket.list(Storage.BlobListOption.prefix(blobNamePrefix));
+        return StreamSupport.stream(blobs.iterateAll().spliterator(), false)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Remove a prefix from a string.
+     */
+    private static String removePrefix(String prefix, String s) {
+        assert s.startsWith(prefix);
+        return s.substring(prefix.length());
+    }
+
+    /**
+     * Convert a list of Blob instances to BlobMetadata instances.
+     */
+    private static List<BlobMetadata> normalizeBlobs(String prefix, List<Blob> blobs) {
+        return blobs.stream()
+            .map(blob -> new PlainBlobMetadata(
+                removePrefix(prefix, blob.getName()), blob.getSize()))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Convert a list of BlobMetadata objects to a Map.
+     */
+    private static Map<String, BlobMetadata> convertBlobsToMap(List<BlobMetadata> blobs) {
+        return blobs.stream().collect(
+            Collectors.toMap(BlobMetadata::name, m -> m));
+    }
```

Comment:
There is a lot of waste going on here.
Most of these methods are only called once, and each is materializing a stream via toList/toMap - if you look at the usages you get a very expensive transformation pipeline.


Instead, this could be changed to something like:

```java
    @Override
    public Map<String, BlobMetadata> listBlobsByPrefix(String blobNamePrefix) {
        String prefix = buildKey(blobNamePrefix);
        return bucket.list(Storage.BlobListOption.prefix(prefix)).streamAll()
            .map(blob -> new PlainBlobMetadata(blob.getName().substring(prefix.length()), blob.getSize()))
            .collect(Collectors.toMap(BlobMetadata::name, m -> m));
    }

```


and

```java
    @Override
    public void delete() {
        // Delete this container and all its contents from the bucket.
        // This means we delete all blobs with the base path as prefix.
        final String basePath = path().buildAsString();
        bucket.list(Storage.BlobListOption.prefix(basePath))
            .streamAll()
            .forEach(Blob::delete);
    }
```